llmclient/llmclient.go

```diff
@@ -0,0 +1,93 @@
+// package llmclient provides a client for the Grafana LLM app.
+// It is used to communicate with LLM providers via the Grafana LLM app
+// using the configuration stored in the app to handle authentication.
+package llmclient
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/sashabaranov/go-openai"
+)
+
+// OpenAI is an interface for talking to OpenAI via the Grafana LLM app.
+// Requests made using this interface will be routed to the OpenAI backend
+// configured in the Grafana LLM app's settings, with authentication handled
+// by the LLM app.
+type OpenAI interface {
+	// Enabled returns true if the Grafana LLM app has been configured for use
+	// with OpenAI.
+	Enabled(ctx context.Context) (bool, error)
+	// ChatCompletions makes a request to the OpenAI Chat Completion API.
+	ChatCompletions(ctx context.Context, req openai.ChatCompletionRequest) (openai.ChatCompletionResponse, error)
+	// ChatCompletionsStream makes a streaming request to the OpenAI Chat Completion API.
+	ChatCompletionsStream(ctx context.Context, req openai.ChatCompletionRequest) (*openai.ChatCompletionStream, error)
+}
+
+type openAI struct {
+	httpClient *http.Client
+	client     *openai.Client
+
+	grafanaURL, grafanaAPIKey string
+}
+
+// NewOpenAI creates a new OpenAI client talking to the Grafana LLM app installed
+// on the given Grafana instance.
+func NewOpenAI(grafanaURL, grafanaAPIKey string) OpenAI {
+	httpClient := &http.Client{}
+	return NewOpenAIWithClient(grafanaURL, grafanaAPIKey, httpClient)
+}
+
+// NewOpenAIWithClient creates a new OpenAI client talking to the Grafana LLM app installed
+// on the given Grafana instance, using the given HTTP client.
+func NewOpenAIWithClient(grafanaURL, grafanaAPIKey string, httpClient *http.Client) OpenAI {
+	url := strings.TrimRight(grafanaURL, "/") + "/api/plugins/grafana-llm-app/resources/openai/v1"
```

Comment:
Could we put part of this path in a const? `/api/plugins/grafana-llm-app`

Reply:
Don't really love the hardcoding here, but loosely held opinion so good to merge as-is 🤷

Reply:
Done (including the `/resources` part) 👍